### ISSUE TITLE
Final_act copy/move semantics is wrong

### ISSUE THREAD

**Issue by ilpropheta** [martinmoene/gsl-lite#19]
Hi Martin, thanks for your great job. 
I think you already know this little issue: I realized that Final_act relies on the copy-elision optimization of the compiler. Although this will be a standard feature of C++17, at the moment it's not. Consider this simple example:

```cpp
auto act = finally([]{cout << "hello\n";});
```
disabling copy-elision (e.g. **-fno-elide-constructors** on clang), you get:

```
hello
hello
hello
```

That is not what you expect. A simple solution consists in just adding a flag (that is also useful in case you provide a "disable-execution" capability).

I blogged about that 4 years ago: 
https://marcoarena.wordpress.com/2012/08/27/mix-raii-and-lambdas-for-deferred-execution/

Microsoft's GSL contains a correct final_act here:
https://github.com/Microsoft/GSL/blob/master/include/gsl_util.h#L60

Keep up the good work!

Marco

**Comment by ilpropheta** [comment]
Little extra: C++17 is going to support [Template argument deduction for class templates](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0091r2.html) so at some point we'll get rid of **finally** factory function, making Final_act just "scoped" (no move/copy semantics at all). This should simplify even more!

**Comment by martinmoene** [comment]
Thanks Marco!

The corrected final_act is available in [Release 0.6.0](https://github.com/martinmoene/gsl-lite/releases/tag/v0.6.0).

Related at Microsoft-GSL:
- Design of Final_act is dependent on copy elision to avoid multiple 'final acts', [Issue #11](https://github.com/Microsoft/GSL/issues/11).
-  Fixing move constructor/semantics of Final_act, [PR #90](https://github.com/Microsoft/GSL/pull/90) .
- Rename Final_act to final_act to follow standard library conventions, [Issue #91](https://github.com/Microsoft/GSL/issues/91).

**Comment by ilpropheta** [comment]
Great! Thanks Martin!
